Clean up UTF-8 conversion code

achristensen07 opened the pull request:
#### efec6c1b6e6da5be8d38c5d3f76b5be71db8f195
<pre>
Clean up UTF-8 conversion code
<a href="https://bugs.webkit.org/show_bug.cgi?id=233612">https://bugs.webkit.org/show_bug.cgi?id=233612</a>

Reviewed by Yusuke Suzuki.

* Source/JavaScriptCore/API/JSStringRef.cpp:
(JSStringGetUTF8CString):
* Source/JavaScriptCore/runtime/ConsoleClient.cpp:
(JSC::ConsoleClient::printConsoleMessageWithArguments):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::putUTF8Triple):
(WTF::StringImpl::utf8ForCharacters):
(WTF::StringImpl::utf8Impl): Deleted.
(WTF::StringImpl::tryGetUTF8ForRange const): Deleted.
(WTF::StringImpl::tryGetUTF8 const): Deleted.
(WTF::StringImpl::utf8 const): Deleted.
(WTF::StringImpl::hashSlowCase const): Deleted.
(WTF::StringImpl::concurrentHash const): Deleted.
* Source/WTF/wtf/text/UTF8ConversionError.h:
* Source/WTF/wtf/text/WTFString.cpp:
* Source/WTF/wtf/text/WTFString.h:
* Source/WTF/wtf/unicode/UTF8Conversion.cpp:
(WTF::Unicode::convertLatin1ToUTF8):
(WTF::Unicode::convertUTF16ToUTF8):
* Source/WTF/wtf/unicode/UTF8Conversion.h:
(): Deleted.
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::utf8Buffer):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::convertUTF16EntityToUTF8):
* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringGetUTF8CStringImpl):

Canonical link: <a href="https://commits.webkit.org/255348@main">https://commits.webkit.org/255348@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f5a4fad2e07ffa63013594b1eb90fc07602c4c7e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91670 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/900 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22270 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101365 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161453 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95675 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/898 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29496 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83984 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97695 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97328 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/527 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78318 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27485 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82451 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82094 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70524 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/83099 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35790 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16121 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/78144 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33544 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17218 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26982 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3707 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37388 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39965 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/80752 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39293 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36335 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17687 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
It seems that this is regressing tryGetUTF8ForCharacters's optimization which avoids allocations. Can you take a look?

Constellation commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -1535,91 +1535,111 @@ size_t StringImpl::sizeInBytes() const
+    char* const bufferEnd = buffer + bufferSize;
+    bool success = convertLatin1ToUTF8(&source, source + length, &buffer, bufferEnd);
+    ASSERT_UNUSED(success, success); // length * 2 should always be large enough for converting from Latin-1.
+    return CString(bufferVector.data(), buffer - bufferVector.data());
```
Here, we are always allocating Vector, but previously, we avoided this when we are ASCII only (see `tryGetUTF8ForCharacters`). Is it expected?

Constellation reviewed (approved):
r=me

Commit:
Clean up UTF-8 conversion code
https://bugs.webkit.org/show_bug.cgi?id=233612

Reviewed by Yusuke Suzuki.

* Source/JavaScriptCore/API/JSStringRef.cpp:
(JSStringGetUTF8CString):
* Source/JavaScriptCore/runtime/ConsoleClient.cpp:
(JSC::ConsoleClient::printConsoleMessageWithArguments):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::putUTF8Triple):
(WTF::StringImpl::utf8ForCharacters):
(WTF::StringImpl::utf8Impl): Deleted.
(WTF::StringImpl::tryGetUTF8ForRange const): Deleted.
(WTF::StringImpl::tryGetUTF8 const): Deleted.
(WTF::StringImpl::utf8 const): Deleted.
(WTF::StringImpl::hashSlowCase const): Deleted.
(WTF::StringImpl::concurrentHash const): Deleted.
* Source/WTF/wtf/text/UTF8ConversionError.h:
* Source/WTF/wtf/text/WTFString.cpp:
* Source/WTF/wtf/text/WTFString.h:
* Source/WTF/wtf/unicode/UTF8Conversion.cpp:
(WTF::Unicode::convertLatin1ToUTF8):
(WTF::Unicode::convertUTF16ToUTF8):
* Source/WTF/wtf/unicode/UTF8Conversion.h:
(): Deleted.
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::utf8Buffer):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::convertUTF16EntityToUTF8):
* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringGetUTF8CStringImpl):

Canonical link: https://commits.webkit.org/255348@main
